Revisit image floats.

jasmussen opened the pull request:
<strike>This adds a wrapping `aside` element to any image block that's floated left or right. From the spec, https://www.w3.org/TR/html52/grouping-content.html#the-figure-element:

> For content that is only tangentially related, or that serves a separate purpose than the surrounding flow, the aside element should be used (and can itself wrap a figure). For example, a pull quote that repeats content from an article would be more appropriate in an aside than in a figure, because it isn’t part of the content, it’s a repetition of the content for the purposes of enticing readers or highlighting key topics.

The above is the reasoning for using the `aside` element to wrap the `figure`. </strike> Per [subsequent discussion](https://github.com/WordPress/gutenberg/pull/7721#issuecomment-408772694), we're going with a div.

But why wrap the `figure` at all?

Because due to issues surfaced in https://github.com/WordPress/gutenberg/pull/7624#issuecomment-402114247, it seems our current implementation isn't responsive.

The challenge is — what if you float a very small image to the left, and write a giant caption. Even if we apply `width: fit-content;` on the `figure`, the caption will expand the `figure` to accommodate as much text as the parent wrapping element will allow. `min-content` doesn't work either, because this will make the `figure` only as wide as the smallest word.

What we have in master works in most cases, through dark magic, but it also only works because we remove the `max-width` from the nested image. This means the image won't resize with the viewport, and is therefore not responsive.

I have explored so many many options for fixing this, and after all this time, what it boils down to is this:

- We can either set a fixed size on floated captions, say 33%, and hope it works for the image that's floated. Not ideal, and you can see the end result here: https://codepen.io/joen/pen/wXbqwN
- We can add a wrapping element around the `figure`, so that it and the `figcaption` can be sized using table rules, as in this PR. See also https://codepen.io/joen/pen/qKGYwy?editors=1100

Why not just set the width on the figure element and float that? Because then we can't accommodate wide images, which rely on an unbounded main column.

It's not ideal that we have to add an extra wrapping element, but it can be semantic, and it feels like the simplest to work with for themers implementing wide images coexisting with floats.

![screen shot 2018-07-05 at 12 09 25](https://user-images.githubusercontent.com/1204802/42317963-f87cbacc-804d-11e8-8849-c5c7a0f8ee28.png)

jasmussen committed:
Revisit image floats.

This adds a wrapping `aside` element to any image block that's floated left or right. From the spec, https://www.w3.org/TR/html52/grouping-content.html#the-figure-element:

> For content that is only tangentially related, or that serves a separate purpose than the surrounding flow, the aside element should be used (and can itself wrap a figure). For example, a pull quote that repeats content from an article would be more appropriate in an aside than in a figure, because it isn’t part of the content, it’s a repetition of the content for the purposes of enticing readers or highlighting key topics.

The above is the reasoning for using the `aside` element to wrap the figure. But why wrap the figure at all?

Because due to issues surfaced in https://github.com/WordPress/gutenberg/pull/7624#issuecomment-402114247, it seems our current implmeentation isn't responsive.

The challenge is — what if you float a very small image to the left, and write a giant caption. Even if we apply `width: fit-content;` on the `figure`, the caption will expand the `figure` to accommodate as much text as the parent wrapping element will allow. `min-content` doesn't work either, because this will make the `figure` only as wide as the smallest word.

What we have in master works in most cases, through dark magic, but it also only works because we remove the `max-width` from the nested image. This means the image won't resize with the viewport, and is therefore not responsive.

I have explored so many many options for fixing this, and after all this time, what it boils down to is this:

- We can either set a fixed size on floated captions, say 33%, and hope it works for the image that's floated. Not ideal, and you can see the end result here: https://codepen.io/joen/pen/wXbqwN
- We can add a wrapping element around the `figure`, so that it and the `figcaption` can be sized using table rules, as in this PR.

Why not just set the width on the figure element and float that? Because then we can't accommodate wide images, which rely on an unbounded main column.

It's not ideal that we have to add an extra wrapping element, but it can be semantic, and it feels like the simplest to work with for themers implementing wide images coexisting with floats.

tofumatt commented on core-blocks/image/index.js:
```diff
@@ -217,6 +217,16 @@ export const settings = {
 			/>
 		);
 
+		if ( 'left' === align || 'right' === align ) {
```
I know it's the style in core and it's unrelated to your patch, but constant-first does my head in… it feels unlike how anyone would speak logically about the conditional. `</rant>`

This looks like it'll need a `deprecated` handler because this changes block markup.

gziolo replied:
We will need a deprecation handler for this change. This will make all old blocks having `align` set to right or left to be recognized as externally modified.

gziolo commented on core-blocks/image/index.js:
```diff
@@ -217,6 +217,16 @@ export const settings = {
+		if ( 'left' === align || 'right' === align ) {
+			return (
+				<aside>
+					<figure className={ classes }>
```
To avoid code duplication, we can assign `figure` to its own variable, similar to what is done for `img` above.

tofumatt replied:
This looks like it could be a variable defined above, so it's not duplicated.

tofumatt reviewed (changes requested):
I think this can be refactored to not repeat the `<figure>` code and needs a deprecation handler, but the idea seems fine to me. I think the extra tag is okay; the explanation is solid, cheers! 👍

tofumatt commented:
@gziolo and I are twins! 😆

jasmussen commented:
Thanks so much for the reviews and examples and thoughts. I think this is a 3.3 milestone thing, in any case it's sort of a sensitive change, which is why I'm going to let it sit for a bit and gather thoughts and opinions. It solves an issue, and I believe it's the best way to solve that issue. But nonetheless I expect there to be valid opinions.

jasmussen commented:
I took a stab at addressing your comments in https://github.com/WordPress/gutenberg/pull/7721/commits/41b3db497c1e678c330a65b3c00f13a64885f3f2, please have a look and see if I did it right.

Also, I realized that this markup change only applies on the frontend. But ideally is reflected in the backend as well. The markup is a bit more verbose in the admin, which hides some of the real world issues themers have to deal with, which is why I didn't catch this the first time round.

In any case, if anyone has bandwidth for helping me with adding the wrapping `aside` element when an alignment is applied, I'd appreciate any input. It seems I can't output just the `<aside>` in the beginning, and then the `</aside>` at the end.

tofumatt commented:
I'm having a look now 👀

SuperGeniusZeb commented:
I am not certain that `<aside>` makes sense for all floated images. `<aside>` should only be used for things that are only tangentially related to the main content and are not actually part of the main content. In many cases, images are a main part of the content even when they are floated, e.g. a diagram used to explain something in a scientific article. I think just `<div>` should be used; no semantics are better than wrong semantics, in my opinion.

jasmussen commented:
I'm back from some vacation ready to look at this. But before I rebase and address semantics feedback (suggestions welcome — div fine?), I'd love some idea on whether we're okay with this. I'm personally very okay and believe it's the only rock solid way to achieve the necessary flexibility for layouts. But I also know it's controversial. @pento any thoughts?

pento commented:
I find myself quite excited about the prospect of using `<div>` for the actual HTML spec reason it exists: as a container element. 🙂

jasmussen commented:
Thanks Gary — is that also a way of stating that you would be okay with the added wrapping element when that element is floated?

pento commented:
It absolutely is my obtuse way of saying I'm okay with adding a wrapping element. 😉

afercia commented:
Aside should be used with caution, and mainly to mark an entire region that is "tangentially related" rather than single pieces of content. The reason is that in ARIA terms, [the aside element is mapped to a complementary role](https://www.w3.org/TR/html-aam-1.0/#details-id-8) which is a landmark role. Landmarks should be used to identify the main regions of a page. Having (potentially) dozens of landmarks in a page would defeat their purpose. (aside: 🙂in core there's the same problem with the widgets).

jasmussen commented:
This PR has been rebased, squashed, updated, tested, to fix the remaining issues. I believe it's ready for final review. The small things:

- aside > div
- works in admin
- feedback addressed

Note that the editor _markup_ differs from the frontend, due to a plethora of editor necessary containers, such as one for resizing and so on. I'm not sure we can ever get the backend markup to match the frontend, so instead I've made sure to map the containers necessary for getting floats to behave the same in the admin as the frontend. See https://github.com/WordPress/gutenberg/pull/7721/commits/d36cce22b89f717185db72fe0425f19223187624 for details.

The trick is that when an image is floated, the parent container (block container) retains its same max-width which usually allows a centered column to coexist with wide images. But it is zeroed out in height. The child element, the figure itself, is then floated, which allows the caption to share the same width as the figure, using a table-caption trick. See also this codepen: https://codepen.io/joen/pen/zLWvrW

**Note that this PR only touches images**. If this is approved and merged in, we'll probably want to backport these changes to other blocks that can float, i.e. video, embed, cover image — any others?.

Screenshots:

![screen shot 2018-08-01 at 10 31 22](https://user-images.githubusercontent.com/1204802/43510989-8ab56b76-9577-11e8-8104-1a2a6a62da4d.png)

![screen shot 2018-08-01 at 10 31 29](https://user-images.githubusercontent.com/1204802/43510992-8cebf6f8-9577-11e8-814f-72fd87b2cff8.png)

![screen shot 2018-08-01 at 10 31 39](https://user-images.githubusercontent.com/1204802/43510993-8ebb353e-9577-11e8-84c9-b41ea5bc8766.png)

![screen shot 2018-08-01 at 10 31 52](https://user-images.githubusercontent.com/1204802/43511000-908cafdc-9577-11e8-9e69-903bef12cd82.png)

jasmussen commented:
Rebased, and added code to account for resized images that have captions. This is now solid, I feel. 

You can test with this dummy markup:

```
<!-- wp:paragraph -->
<p>Big image</p>
<!-- /wp:paragraph -->

<!-- wp:image {"id":37} -->
<figure class="wp-block-image"><img src="http://gutenberg/wp-content/uploads/2018/04/pic9.jpg" alt="" class="wp-image-37"/><figcaption>Loem ipsum dolor sit amet, ferri vidisse nam eu, ad nec copiosae mnesarchum vituperatoribus. Te brute dicunt sea, ut vis omnium menandri, ut sumo aliquam has. Eum aperiam interpretaris at, sea et recusabo expetenda, omnis tibique mea no. Pri suas partem ea, ius sonet numquam offendit cu, ad simul admodum pri. Eum cu unum choro albucius.</figcaption></figure>
<!-- /wp:image -->

<!-- wp:paragraph -->
<p>Small image</p>
<!-- /wp:paragraph -->

<!-- wp:image {"id":36,"width":248,"height":164} -->
<figure class="wp-block-image is-resized"><img src="http://gutenberg/wp-content/uploads/2018/04/pic8.jpg" alt="" class="wp-image-36" width="248" height="164"/><figcaption>Loem ipsum dolor sit amet, ferri vidisse nam eu, ad nec copiosae mnesarchum vituperatoribuas. Eum aperiam interpretaris at, sea et recusabo expetenda, omnis tibique mea no. Pri suas partem ea, ius sonet numquam offendit cu, ad simul admodum pri. Eum cu unum choro albucius.</figcaption></figure>
<!-- /wp:image -->

<!-- wp:paragraph -->
<p>Here's a float</p>
<!-- /wp:paragraph -->

<!-- wp:image {"id":32,"align":"right","width":297,"height":197,"className":"alignright is-resized"} -->
<div class="wp-block-image alignright is-resized"><figure class="alignright is-resized"><img src="http://gutenberg/wp-content/uploads/2018/04/pic4.jpg" alt="" class="wp-image-32" width="297" height="197"/><figcaption>Loem ipsum dolor sit amet, ferri vidisse nam eu, ad nec copiosae mnesarchum vituperatoribus. Te brute dicunt sea, ut vis omnium menandri, ut sumo aliquam has. Eum aperiam interpretaris at, sea et recusabo expetenda, omnis tibique mea no. Pri suas partem ea, ius sonet numquam offendit cu, ad simul admodum pri. Eum cu unum choro albucius.</figcaption></figure></div>
<!-- /wp:image -->

<!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet, ferri vidisse nam eu, ad nec copiosae mnesarchum vituperatoribus. Te brute dicunt sea, ut vis omnium menandri, ut sumo aliquam has. Eum aperiam interpretaris at, sea et recusabo expetenda, omnis tibique mea no. Pri suas partem ea, ius sonet numquam offendit cu, ad simul admodum pri. Eum cu unum choro albucius.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Eu integre accusata prodesset est, sed te impetus gubergren conceptam, ex sed wisi nostrum ocurreret. Esse velit omittantur ius te, alii dissentias ei vis. At sed unum veritus fabellas. Te volutpat appellantur duo. Ad natum fuisset intellegebat eam, causae invidunt usu id, et vis impetus appetere.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Ea veniam homero eam. Ex inimicus molestiae cum, debet scaevola at eos. Vis assum veritus ut, has ea nostrud accusata, offendit appareat comprehensam ea pro. Ad quo quem veritus appellantur, te est quas phaedrum, eum alia habeo ad. Ei est erroribus imperdiet, omnis dicam propriae sed no. His vitae oratio fierent ne, cu duo tota eligendi, electram rationibus in qui.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Est quis reque cetero ad. Sea id autem nominavi deseruisse. Veniam qualisque definitionem pri id, ea autem feugiat delenit ius, mei at lorem affert accumsan. Dicat eruditi cu est, te pro dicant pericula conclusionemque, ei vim detracto euripidis intellegam. Eius postea volumus mei ad.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Prima ridens denique his te, ferri illum volumus an his. Eu vel dicat homero qualisque, vitae regione deserunt vis ei. Graeci incorrupte liberavisse no mea, saepe voluptaria usu ex, vis dicant euismod id. At dolor reprimique eos, quo altera detraxit moderatius id. Quo iudico utinam eu, ad alia munere mel.</p>
<!-- /wp:paragraph -->
```

jasmussen commented:
Rebased, squashed, fixed a few issues. I think this is solid. It uses [this method](https://codepen.io/joen/pen/zLWvrW), and it is responsive (thanks to @sarahmonster), and figcaptions now behave as intended in every situation — floated, wide, fullwide, floated and resized, not floated but resized, centered, etc.

Backend:

<img width="793" alt="screen shot 2018-08-06 at 10 27 11" src="https://user-images.githubusercontent.com/1204802/43705677-7f0f0d50-9963-11e8-97f7-91936987e38d.png">

Frontend in https://github.com/jasmussen/navi theme:

<img width="883" alt="screen shot 2018-08-06 at 10 27 18" src="https://user-images.githubusercontent.com/1204802/43705684-8711ca60-9963-11e8-9b5f-83482d87b293.png">

Let's test this and get it in. Thank for your reviews.

SuperGeniusZeb commented:
I noticed that when inserting an image, it defaults to being 100% of the width of the block, regardless of the size of the image. Previously, a small image would only be its normal width by default (unless it was larger than the block, it which case it would be the width of the block). Is this an intentional/desirable change in behavior?

jasmussen commented:
Alright, I think I fixed the issue with smaller images.

sarahmonster commented on core-blocks/image/index.js:
```diff
@@ -217,15 +218,59 @@ export const settings = {
-		return (
-			<figure className={ classes }>
+		const figure = (
+			<Fragment>
```
Why not use `<React.Fragment>` instead of importing a package specifically for this purpose?

tofumatt replied:
Huh, nice catch. I'm guessing our own `Fragment` predates the React one, but we should look into ditching ours.

aduth replied:
No, this is correctly implemented as-is. We don't and shouldn't use React directly.

tofumatt replied:
Ah, I guess that's so we can control the API surface better?

sarahmonster reviewed (approved):
I tested this and didn't manage to break anything! 🎉 Left you one minor nitpicky comment that may or may not be relevant, so do with it as you will. (I'm not sure about how the WordPress packages work, exactly.)

kjellr commented:
> Alright, I think I fixed the issue with smaller images.

@jasmussen: I'm seeing some oddities with the resize handles on smaller images. They aren't adhering to the image by default: 

![screen shot 2018-08-07 at 10 49 28 am](https://user-images.githubusercontent.com/1202812/43783561-0503b006-9a30-11e8-829f-0d368e6b22f6.png)

![screen shot 2018-08-07 at 10 51 14 am](https://user-images.githubusercontent.com/1202812/43783570-086030bc-9a30-11e8-9be6-aa9381354c80.png)

![screen shot 2018-08-07 at 10 50 20 am](https://user-images.githubusercontent.com/1202812/43783574-0bbf7ea2-9a30-11e8-8286-7a4f00e42257.png)

If I drag them around a bit, they fix themselves, but those screenshots depict the way they look when first inserted, in the latest versions of Chrome + Safari. These are the images I used, in case you want to try & reproduce: https://cloudup.com/c-S3iSGQkok

jasmussen commented:
Thanks a bunch, @kjellr, solid catch. I think that issue was in master also, but was hidden because a small non resized image was just huge. Maybe?

In any case I believe I pushed a fix in https://github.com/WordPress/gutenberg/pull/7721/commits/8e8848fd45387f748bce673d29179234f5453cb0

![resizer](https://user-images.githubusercontent.com/1204802/43894627-05dfe928-9bd3-11e8-9d23-e7dd173c9517.gif)

But this is a fix that sort of can have repercussions. I've tested it well myself and I think it's good, but I'd appreciate thorough tests if you have time. Thanks.

kjellr commented:
Thanks, @jasmussen! That's looking much better. I'm still seeing the caption text center aligned in the entire box though:

<img width="580" alt="screen shot 2018-08-09 at 9 56 51 am" src="https://user-images.githubusercontent.com/1202812/43903916-782d03d4-9bbb-11e8-99b8-6c62e6858003.png">

After resizing the handles slightly, the caption snaps over to its intended placement though. 🤔 

Also, I am seeing a more prominent height issue with big images now:

![image](https://user-images.githubusercontent.com/1202812/43904028-bf9bae64-9bbb-11e8-8e50-1d6c7576b023.gif)

I'm able to reproduce that in both Safari and Chrome.

jasmussen commented:
Solid testing yet again, @kjellr. I think I fixed it:

![fixed](https://user-images.githubusercontent.com/1204802/43946633-0e2e525a-9c86-11e8-9d8a-5089f421f43d.gif)

But because the fix includes making the figure behave as a table in even more cases than before, I'd love a thorough round of testing yet again. Thanks.

kjellr commented:
@jasmussen: So close! Everything looks 💯  except I think some of these styles extended out into the placeholder by mistake: 

<img width="644" alt="screen shot 2018-08-10 at 11 28 15 am" src="https://user-images.githubusercontent.com/1202812/43966857-fb3d823c-9c90-11e8-8383-a7a2d896cfbb.png">

The rest of the placeholders look just fine, and besides this I think we're good to go. 👍

jasmussen commented:
Fixed it Kjell, thanks a bunch for the solid testing! 

<img width="822" alt="screen shot 2018-08-13 at 09 47 16" src="https://user-images.githubusercontent.com/1204802/44019028-ee56ed3e-9edd-11e8-8d95-5097ed3b027f.png">

jasmussen commented:
Rebased, tested again, and fixed an issue in https://github.com/WordPress/gutenberg/pull/7721/commits/5af3733efcdd7604df2708fec24be09807a06bb5 that added the table styles twice to floats and centered images. We use these table style to ensure captions are no wider than the image itself. 

This should work now, be solid, and still be good to ship asap. Good to test:

- Normal, wide, fullwide, floated left, centered, right images. Test in editor, and in frontend, though the theme might not take full advantage of wide images, but floats with captions that fit should still work.
- Try floats and centered images that are small, smaller at least than the minimum contentwidth.
- Try with long captions in all cases. Try without captions.
- Try resizing images, with and without captions.

All of those are solid for me.

tofumatt committed:
chore: tweak documentation

tofumatt commented on packages/block-library/src/image/index.js:
```diff
@@ -216,15 +217,59 @@ export const settings = {
+				return (
+					<figure className={ classes }>
+						{ href ? <a href={ href }>{ image }</a> : image }
+						{ caption && caption.length > 0 && <RichText.Content tagName="figcaption" value={ caption } /> }
```
Does this need to be `caption.length > 0` instead of just `caption.length`? 🤷‍♂️

tofumatt committed:
docs: Tweak some comments

Add clarity to HTML elemeents in the comments which might also be read as words.

tofumatt commented on packages/editor/src/components/block-list/style.scss:
```diff
@@ -346,14 +346,16 @@
 				content: none;
 			}
+
+			// This margin won't collapse on its own, so zero it out.
```
I'm not *quite* sure what this means in this context, so maybe a bit more commenting would be handy.

tofumatt reviewed (approved):
Largely looks good to me. I noticed that I could get certain widths to float in the editor but not my theme, but maybe that's just gonna be how it is?

<img width="888" alt="screenshot 2018-08-16 15 34 34" src="https://user-images.githubusercontent.com/90871/44215265-00fd3680-a16a-11e8-8567-1da058aad667.png">
vs
<img width="738" alt="screenshot 2018-08-16 15 34 29" src="https://user-images.githubusercontent.com/90871/44215266-00fd3680-a16a-11e8-9c01-d686a327c986.png">

But it largely seems to work well with/without captions.

The code seems solid. I made a comment about one thing I thought could use more clarification but otherwise looks good to me.

tofumatt commented:
I think this is good to ship but I'm moving it to `3.7` because I think it could use a bit more testing by others to verify it's good-to-go… it's already been without a merge for awhile so I'm not convinced it's urgent and also there's a failing test. But if anyone else looks at it and says it's good **and** the test is fixed, feel free to move it back to `3.6` if there's time 😄

tofumatt commented:
Looks like `test/integration/full-content/full-content.spec.js` is the failure, but I'm about to run out to squash and am not sure why the failure is happening. I can take a look when I'm back online in a few hours though.

tofumatt commented:
I just noticed an issue with floated images and I'm not sure it's related to this change or not. Here's a screencast:

![2018-08-16 20 54 28](https://user-images.githubusercontent.com/90871/44231666-c1990f00-a196-11e8-9cfc-2e0fcfd57bb6.gif)

When clicking on a floated image, it selects both the image and block before it. I need to double-click on the image.

If that's not a regression okay, but wanted to check here in case it was.

tofumatt commented:
Looking at the test now…

ZebulanStanphill commented:
@tofumatt Unfortunately, that selection behavior is unrelated, and has been happening in `master` for a long time. It only happens in certain situations, but I have not bothered to determine which ones. It has been that way since I started using Gutenberg around 1.6 or earlier. It has probably been there since float alignment was added to Gutenberg.

tofumatt commented:
Looks like #7500 and #8293 cover the issue I ran into with floated image click behaviour. Coolio.

tofumatt committed:
fix: Update test fixtures for new image float

tofumatt commented:
I just pushed an update that changes the fixtures and passes all tests, but I'm dubious as to why that fix works. The `<div>` in the code isn't setting a `className`, so why are the fixtures demanding one? Something is up requiring 9b973e7de6462785f1eceb7d339c5b60a39c1c9d to pass tests, could someone else have a look because this doesn't make sense to me.

tofumatt reviewed (changes requested):
After 9b973e7de6462785f1eceb7d339c5b60a39c1c9d is needed to pass I think this needs another set of eyes. Maybe I'm missing something obvious?

youknowriad commented:
I haven't looked at the code but from my testing, there's one difference with master that I think master is getting it right but would want your thoughts

master:

<img width="646" alt="screen shot 2018-08-17 at 11 52 40" src="https://user-images.githubusercontent.com/272444/44262757-1333af00-a214-11e8-975b-e0c4b3324dce.png">

This branch:

<img width="655" alt="screen shot 2018-08-17 at 11 52 21" src="https://user-images.githubusercontent.com/272444/44262782-234b8e80-a214-11e8-8938-c50d08fbd75e.png">


---
Basically when you float an image in master that is followed by another image, the second one appears under the floated image (because it has a width: 100% I guess) but in the current branch, the second one appears in the remaining area next to the floated image. (you could argue that this is the 100% remaining area)

So I'm not really certain 100% what's the right thing, maybe master but it's not obvious.

youknowriad commented on packages/block-library/src/image/edit.js:
```diff
@@ -432,6 +432,7 @@ class ImageEdit extends Component {
 								<Fragment>
 									{ getInspectorControls( imageWidth, imageHeight ) }
 									<ResizableBox
+										className="wp-block-image__resizer"
```
This should be `block-library-image__edit-resizer` if we want to be strict about our className guidelines :)

youknowriad replied:
yes, I think so the `wp-block-*` convention is only for styles that can be used on the frontend, which I think is not the case here.

youknowriad commented on packages/block-library/src/image/editor.scss:
```diff
@@ -103,3 +106,43 @@
+// is floated instead of the block itself.
+[data-type="core/image"][data-align="center"],
+[data-type="core/image"][data-align="left"],
+[data-type="core/image"][data-align="right"] {
```
This feels like a specific fix to a generic problem, how is this solved for other floated blocks? Can't we have a generic solution to left/right aligned blocks?

youknowriad commented on packages/block-library/src/image/index.js:
```diff
@@ -216,15 +217,59 @@ export const settings = {
+
+		if ( 'left' === align || 'right' === align || 'center' === align ) {
+			return (
+				<div>
```
Wasn't this meant to be an aside (from the PR description)?

jasmussen replied:
Yes, but per discussion in https://github.com/WordPress/gutenberg/pull/7721#issuecomment-408772694, I reverted to using a div. I will update the description.

tofumatt replied:
I'm also curious about this though, because it doesn't have a `className` attribute yet in the fixtures the tests expect/output a `<div>` with a `class` attribute. See: 66f6f01a70bd725b3bcada40776156fa86b0cb9d

I'm a bit confused by that...

jasmussen commented:
I can reproduce the issue with the 2nd float, taking a look.

jasmussen committed:
Fix resize handle names

jasmussen committed:
Revert the part that limits caption width on unresized images

We have a plethora of different combinations of images and captions.

Floated images with captions, non floated images with captions, small images with captions, etc. etc.

In general we use a `display: table;` trick to get the caption to size itself according to the image. However if an unfloated image follows a floated image, this means the unfloated image can size itself down to "fit in the available space", which is not what we want. So for now, if you upload a small image and _don't_ float it, a wide caption will be as centered as the block itself.

This might be worth revisiting, but at this point I'd like to look at that separately from this PR which is already big.

jasmussen commented:
Okay, I've rebased a bit, squashed a bit, addressed feedback about CSS classnames, and I have a fix for the float issue caught in https://github.com/WordPress/gutenberg/pull/7721#issuecomment-413830182. Excellent, excellent catch, Riad. 

So the reason that happened was interesting. The thing is, we have a plethora of different combinations of images and captions: floated images with captions, non floated images with captions, small images with captions, resized and floated, resized but not floated, etc. etc.

In general we use a `display: table;` trick to get the caption to size itself according to the image. This works because of the unique rules of the table. However if an unfloated image follows a floated image, this means the unfloated image can size itself down to "fit in the available space", which is not what we want. 

For now, I've removed those table rules for *images that are not resized, and not floated*. Which means you could get this:

![screen shot 2018-08-17 at 13 57 35](https://user-images.githubusercontent.com/1204802/44265427-bee5fa80-a226-11e8-8630-ab73801a3eaa.png)

That's an image block with a small image and a big caption. Because it doesn't have an explicit size, applying table rules and a float to that would cause the situation Riad mentioned where it would scale down to fit available space next to a float.

We _could_ revisit this and try and improve it so the caption is no wider than the image, even if not floated, but I would prefer to do this separately so as to not make this PR any more complex. And even then, there are vast caption improvements in this PR, in addition to the float improvements. These are worth getting in. And honestly, if you want to upload a small image, that isn't centered or floated, and you want to add a giant caption, hey, perhaps it's okay it's not a super tailored experience?

jasmussen commented:
Thanks for all the reviews. Once the dust settles from 3.6, I'd like to get this in.

jasmussen commented:
Okay, I'm merging now. While it's a big change, it's been in progress for a while, it's solid, and it's fixing a number of issues. Thanks for the reviews everyone!

jcklpe commented:
@jasmussen Glad to have found this! I've been struggling to find the answer to this question and it's popped sporadically and completely threw off a legacy theme I had developed midway through Gutenberg being released and so it's good to now actually get a concrete bit of information about it. 

It has recently popped up again as I am now working on developing a theme for the HeadlessWP+React+SSR framework [frontity](https://frontity.org). 

My latest issue I posted in response to that problem. 
https://github.com/WordPress/gutenberg/issues/18122 

I defer to your expertise in this stuff but I'd like to ask a quick question and offer some feedback:

First
"Why not just set the width on the figure element and float that? Because then we can't accommodate wide images, which rely on an unbounded main column."

I'm not sure I totally understand this reasoning. Why would people apply a float to a wide image? Aren't wide images full width etc? So they would wouldn't really be floated right? 

Second if a wrapper div is needed, I think it should be consistent across all images.   I find it kind of confusing that images are div->figure and then wide images etc are just figure. Consistency in the structure would be nice because there is some CSS that applies to all images that I would like to be able to target, and then could have the variations targeted in a nested fashion (via css-in-js or scss etc) like so:
```css
div.wp-image-block. {
opacity: 50%;
background-color: yellow;
   &.alignleft {
     float: left
   }

   &.fullwidth {
      ...
   }
}
```

Preferable to me though is just having figure elements but I understand that you guys have already given a lot of thought to this stuff. (I'm also curious how much of some of these problems could be addressed in a flexbox fashion? )

jasmussen commented:
@jcklpe Sorry for the headaches this has caused you!

Let me try and answer to the best of my ability!

> > "Why not just set the width on the figure element and float that? Because then we can't accommodate wide images, which rely on an unbounded main column."
> 
> I'm not sure I totally understand this reasoning. Why would people apply a float to a wide image? Aren't wide images full width etc? So they would wouldn't really be floated right?

Yes that phrasing was obtuse. You are correct, a wide and fullwide image will never be floated, so that's not the problem.

The problem is: how do you create a layout that accommodates both wide, fullwide, and floated images? [This codepen](https://codepen.io/joen/pen/zLWvrW) shows how the block editor solves it. 

Traditionally in themes you'd have a centered main column, something a la `.main { max-width: 800px; margin: 0 auto; }`. Anything you'd float inside there would be nicely flush with text as well. But this layout makes it hard to add wide and fullwide images and blocks — you can use negative margins, or `vw` units and make it work. But this does not work inside the editor because there you have to account for the WordPress navigation and the editor sidebar.

Therefore the approach taken for the editor styles, is one where there isn't a `max-width` on the main column, instead there are max-widths on every element that goes inside, a la this: 

```
.site-content > *:not( .alignwide ):not( .alignfull ) {
  max-width: 50%;   /* Could be a pixel value too: this is the one that decides the width of the text */
  margin-left: auto;
  margin-right: auto;
}
```

This allows wide and fullwide images to go from edge to edge without having to use `vw` units or negative margins, and it constrains blocks to a centered main column.

Floats, as you can imagine, become a challenge, because if you simply float a block, it goes right up against the edge of the viewport, where usually you'd want it to go up against the margin of the perceived main column. 

We accomplish this by having a wrapping container element, the div, which allows you to assign a  max-width to that container, and then float the figure inside. 

> Second if a wrapper div is needed, I think it should be consistent across all images. I find it kind of confusing that images are div->figure and then wide images etc are just figure. Consistency in the structure would be nice because there is some CSS that applies to all images that I would like to be able to target, and then could have the variations targeted in a nested fashion (via css-in-js or scss etc) like so:

Yes indeed, it was unfortunate the change had to be made as such. The truth is, _any image inserted after the change was made, will have the wrapping div and be consistent going forward_. But images inserted before that, don't break because they lack the wrapper. You could re-insert the old images, and they will be updated to the new markup. 

It's a headache, I understand, but ultimately the wrapping div was added in order to give more control to themers like yourself. Wide images, floats, figures that are only as wide as the caption inside, and combinations of the three present quite the challenge to solve. 

Hope this helps!

jcklpe commented:
@jasmussen  yes thank you that does help. 

Quick question though, for the purposes of my developer self education. I've been until now applying my elements to the perceived main column via margin similar to how you do here:

```
.site-content > *:not( .alignwide ):not( .alignfull ) {
  max-width: 50%;   /* Could be a pixel value too: this is the one that decides the width of the text */
  margin-left: auto;
  margin-right: auto;
}
```
What I do instead is something more like this though:

```
.site-content > *:not( .alignwide ):not( .alignfull ) {
margin: 25px 20vw 0;
}
```

and then when something is floated right I just do this

```
.alignright {
float: right;
margin: 25px 20vw 25px 25px;
}
```
Why not do something like this second example? I see you mention vw being an issue for the editor but you seem to be saying that in the context of full width images right? Or were you saying that this is an issue for floated images in the editor because vw is based on viewport width and in the editor that layout won't be taking into account the more limited space due to the editor and wordpress admin bar etc? Why not simply have separate rules output for the admin versus the frontend? isn't that what is already done in wordpress themes? One can just build separate build outputs based on scss imports right?

jasmussen commented:
> Why not do something like this second example? I see you mention vw being an issue for the editor but you seem to be saying that in the context of full width images right? Or were you saying that this is an issue for floated images in the editor because vw is based on viewport width and in the editor that layout won't be taking into account the more limited space due to the editor and wordpress admin bar etc? 

Yes, the last part exactly. 

![Screenshot 2019-11-01 at 08 31 05](https://user-images.githubusercontent.com/1204802/68009404-2d737480-fc82-11e9-9f6b-ec7cd39a5108.png)

Consider here that 20vw to the right and left of the paragraph below the image would make for a very thin paragraph. 

_However_ you can definitely use vw units for the front-end. What we have learned from user test is that if the editor looks like the frontend, usability is drastically increased, which is why editor styles are very much recommended. However they don't have to be 100% accurate to the front-end. So if you use % in the editor and it _looks_ good, that's totally fine.

A personal goal of mine is to make it so you can load the theme CSS directly into the editor without changes. This is a bit difficult, but there are ways to accomplish that, but it's going to take some time to make this possible. 

> At the very least would it be possible to add some kind of class to the top level wp-image-block wrappers that contain an alignedright figure? That way when people want to use vw they can just apply a float to the top level container block and write a second set of styles for their admin view.

This layout:

![Screenshot 2019-11-01 at 08 41 54](https://user-images.githubusercontent.com/1204802/68009937-7ed03380-fc83-11e9-817b-a5f4fc8dadf1.png)

brings with it this markup:

```
<figure class="alignright size-large is-resized"><img src="..." alt="" class="wp-image-149" width="287" height="192">
    <figcaption>Loem ipsum dolor sit amet, ferri vidisse nam eu, ad nec copiosae mnesarchum vituperatoribus. Te brute dicunt sea, ut vis omnium menandri, ut sumo aliquam has. Eum aperiam interpretaris at, sea et recusabo expetenda, omnis tibique mea no. Pri suas partem ea, ius sonet numquam offendit cu, ad simul admodum pri. Eum cu unum choro albucius.</figcaption>
</figure>
```

Was this helpful?

jcklpe commented:
@jasmussen 
> _However_ you can definitely use vw units for the front-end. What we have learned from user test is that if the editor looks like the frontend, usability is drastically increased, which is why editor styles are very much recommended. However they don't have to be 100% accurate to the front-end. So if you use % in the editor and it _looks_ good, that's totally fine.
> 
> A personal goal of mine is to make it so you can load the theme CSS directly into the editor without changes. This is a bit difficult, but there are ways to accomplish that, but it's going to take some time to make this possible.

That makes sense. Personally I think it's so easy to break out variations in editor/admin styles and frontend styles, and it's so difficult to get files matching 100% between both, and inevitably that stuff is going to get changed by themes anyway, that it seems like preserving markup is more important. 

That's just my personal perspective though and a lot of it is informed by the fact that I'm working with WP as a headless CMS so the styling difference between front and backend doesn't even really make a difference to me. I understand that ya'll are operating in the context of the organization and what ya'll feel is it's best interests. 

> > At the very least would it be possible to add some kind of class to the top level wp-image-block wrappers that contain an alignedright figure? That way when people want to use vw they can just apply a float to the top level container block and write a second set of styles for their admin view.
> 
> This layout:
> 
> ![Screenshot 2019-11-01 at 08 41 54](https://user-images.githubusercontent.com/1204802/68009937-7ed03380-fc83-11e9-817b-a5f4fc8dadf1.png)
> 
> brings with it this markup:
> 
> ```
> <figure class="alignright size-large is-resized"><img src="..." alt="" class="wp-image-149" width="287" height="192">
>     <figcaption>Loem ipsum dolor sit amet, ferri vidisse nam eu, ad nec copiosae mnesarchum vituperatoribus. Te brute dicunt sea, ut vis omnium menandri, ut sumo aliquam has. Eum aperiam interpretaris at, sea et recusabo expetenda, omnis tibique mea no. Pri suas partem ea, ius sonet numquam offendit cu, ad simul admodum pri. Eum cu unum choro albucius.</figcaption>
> </figure>
> ```
> 
> Was this helpful?

That's weird because I don't get that markup: 

![image](https://user-images.githubusercontent.com/9451726/68041148-e98e6880-fc9d-11e9-8eaf-2e965537245e.png)
That's a fresh default install of WP with twentynineteen installed. 

What I'm asking is if instead of this:

```
<div class="wp-block-image">
	<figure class="alignright">
		<img src="url" alt="iFly Virtual Reality" data-pagespeed-url-hash="1116608801" onload="pagespeed.CriticalImages.checkImageForCriticality(this);" googl="true">
	</figure>
</div>
```

we could have this:

```
<div class="wp-block-image block-alignright">
	<figure class="alignright">
		<img src="url" alt="iFly Virtual Reality" data-pagespeed-url-hash="1116608801" onload="pagespeed.CriticalImages.checkImageForCriticality(this);" googl="true">
	</figure>
</div>
```

That way if a theme dev is using vw then they can more easily target the wp-image-blocks that contain alignright figures. 

because as it is now, if the paragraphs etc and images are all laid out using vw, it's kind of hard to cleanly target images that have been aligned right. The framework I'm using uses css-in-js (specifically the emotion framework) which applies nesting rules in such a way that that I can successfully target wp-image-blocks that contain an alignright figure but only by using css unset styling rules. It results in very unclean styling logic. (I might also need to refactor my code some too but I'm pretty sure that I am in fact running into this problem in the way that I think I am. )

jasmussen commented:
Indeed, the markup you get is what I get as well. I'm not very confused how I got the other markup to show up. A glitch in the matrix? A case of lack of sleep on my part? My bad, in any case. 

I do understand how if the alignment class was on the parent node, there would be more control on the part of themers, yes. I think you should feel free to open a ticket to suggest that additional class.

jcklpe commented:
@jasmussen  Cool. I've opened a request here: https://github.com/WordPress/gutenberg/issues/18276

feastdesignco commented:
Just wanted to drop a comment here that randomly deciding to display images like tables (only when the is-resized class is applied) is completely insane and causes a bunch of unexpected headaches. 

This is extremely disappointing.

ZebulanStanphill commented:
It would be helpful if you could explain why it's bad and what would be better. Maybe open an issue suggesting how to fix it?
